### PR TITLE
docs(repo-scan): repair pinned installation flow

### LINKED ISSUE #2774: repo-scan: documented install fails verbatim (both git commands), and the skill ships no implementation
Verified against `origin/main` @ `569b1d5b` (2026-08-12), read through `gh api repos/affaan-m/ecc/contents/...` rather than a local checkout — deliberately, because a stale-checkout claim is what sank #2599.

## 1. The documented install does not work

`skills/repo-scan/SKILL.md:20-30` gives this block. Two of its commands fail outright.

```bash
mkdir -p ~/.claude/skills/repo-scan
git init repo-scan
cd repo-scan
git remote add origin https://github.com/haibindev/repo-scan.git
git fetch --depth 1 origin 2742664      # ← fails
git checkout --detach FETCH_HEAD        # ← fails
cp -r . ~/.claude/skills/repo-scan      # ← copies .git
```

Run as written:

```console
$ git fetch --depth 1 origin 2742664
fatal: couldn't find remote ref 2742664

$ git checkout --detach FETCH_HEAD
fatal: git checkout: --detach does not take a path argument 'FETCH_HEAD'
```

- **`git fetch <sha>`** — GitHub rejects the raw SHA as a remote ref for this repository. The commit is genuine and reachable: after a full clone, `git cat-file -t 2742664` returns `commit`, and it is the second-newest commit on `master`. So the pin is valid; only the fetch-by-SHA mechanism is unsupported here.
- **`git checkout --detach FETCH_HEAD`** — `--detach` does not accept an argument. `git checkout --detach` alone, or `git checkout FETCH_HEAD`, are the valid forms.
- **`cp -r .`** — copies `.git` into the installed skill. 4.8 MB installed vs 1.6 MB with history dropped.

Because the two failures are sequential, a user following the README verbatim never reaches a working install.

## 2. Verified working replacement

This is what I actually ran to install it successfully:

```bash
git clone https://github.com/haibindev/repo-scan.git /tmp/repo-scan-src
cd /tmp/repo-scan-src
git checkout 2742664        # the pinned commit, reachable after a full clone
rm -rf .git                 # drop history before installing
mkdir -p ~/.claude/skills/repo-scan
cp -r . ~/.claude/skills/repo-scan/
```

If shallow-fetch is wanted for its own sake, `git fetch --depth 1 origin master && git checkout 2742664` will not work either — the pinned commit is not in a depth-1 fetch. A full clone is the straightforward fix.

## 3. Present in all three shipped locales

Same broken block, all at `569b1d5b`:

- `skills/repo-scan/SKILL.md:27-28`
- `docs/zh-CN/skills/repo-scan/SKILL.md`
- `docs/ja-JP/skills/repo-scan/SKILL.md`

## 4. Minor: the install path assumes a default config dir

`~/.claude/skills` is only the right target when `CLAUDE_CONFIG_DIR` is unset. On my machine (multiple config dirs) I installed there and the running session could not see the skill at all — its `skills` entry resolved elsewhere. `readlink "${CLAUDE_CONFIG_DIR:-$HOME/.claude}/skills"` is a reliable way to name the correct destination. Raising this as a docs suggestion, not a defect claim; I have only verified it on my own setup.

## 5. Question: should this skill describe itself as an installer?

`skills/repo-scan/` at `main` contains exactly one file:

```console
$ gh api repos/affaan-m/ecc/contents/skills/repo-scan --jq '.[] | .type + "  " + .name'
file  SKILL.md
```

No `commands/repo-scan*`, no `agents/repo-scan*`. The frontmatter reads:

> description: Cross-stack source code asset audit — classifies every file, detects embedded third-party libraries, and delivers actionable four-level verdicts per module with interactive HTML reports.

The skill does not do any of that. It documents how to install `haibindev/repo-scan`, which does. Since the description is what the model matches on when routing, invoking `/ecc:repo-scan` surfaces install instructions at the moment a user asked for an audit.

I am not arguing pointer-skills are wrong — `origin: community` makes the provenance clear, and the tool it points to is good. The question is whether their `description` should say what the skill *delivers* (install instructions for X) rather than what X delivers, so routing and expectations line up. Happy to send a PR for either the install-block fix alone or both, whichever you prefer.

## Environment

- ECC `main` @ `569b1d5b`
- `git version 2.50.1 (Apple Git-155)`, macOS 26.5
- Target tool: `haibindev/repo-scan` @ `2742664`

### PR DESCRIPTION
**Pull request by dajiaohuang**
## What Changed

- replace the unfetchable short-SHA installation flow with a full 40-character pinned commit
- clone into a temporary review directory before installation
- install only tracked files through `git archive`, avoiding copied `.git` metadata
- honor `CLAUDE_CONFIG_DIR`
- keep English, Simplified Chinese, and Japanese installation commands synchronized
- add a regression test for all three documentation surfaces

## Why This Change

The documented `git fetch --depth 1 origin 2742664` command treats a short object ID as a remote ref and fails when followed verbatim. The subsequent `cp -r .` would also copy repository metadata into the installed skill.

Closes #2774

## Testing Done

- [x] Manual testing completed
- [x] Automated focused tests pass locally
- [x] Edge cases considered and tested

Validation performed:

- exact Git Bash install flow checked out `2742664ebcad1450c208eda0ae45d3c17fad5dd8`
- archive extraction produced 29 tracked files and no `.git` directory
- `node tests/skills/repo-scan-install.test.js`
- `node scripts/ci/validate-skills.js`
- `npm run catalog:check`
- `node scripts/ci/validate-no-personal-paths.js`
- `npx markdownlint` on all three changed skill documents

## Type of Change

- [ ] `fix:` Bug fix
- [ ] `feat:` New feature
- [ ] `refactor:` Code refactoring
- [x] `docs:` Documentation
- [x] `test:` Tests
- [ ] `chore:` Maintenance/tooling
- [ ] `ci:` CI/CD changes

## Security & Quality Checklist

- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly
- [ ] Shell scripts pass shellcheck (if applicable)
- [x] Pre-commit hooks pass locally (if configured)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## Documentation

- [x] Updated relevant documentation
- [x] Added comments for complex logic
- [ ] README updated (if needed)

### CODE REVIEW

File: tests/skills/repo-scan-install.test.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  assert.ok(block.includes('${CLAUDE_CONFIG_DIR:-$HOME/.claude}'), `${relativePath} must honor CLAUDE_CONFIG_DIR`);
+  assert.ok(!block.includes('cp -r .'), `${relativePath} must not copy .git metadata`);
+  assert.ok(!block.includes('git fetch --depth 1 origin 2742664\n'), `${relativePath} must not fetch the short SHA`);
+}
+
+for (const block of blocks.slice(1)) {
+  assert.strictEqual(block, blocks[0], 'translated installation commands must stay synchronized');
+}
+
+console.log(`  PASS ${skillFiles.length} installation blocks use the verified pinned flow`);
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/skills/repo-scan-install.test.js]: Fixed in 53408f7b. The regression test now emits the repository aggregate runner format with Passed: 3 and Failed: 0; the focused test and validation checks pass.

File: skills/repo-scan/SKILL.md [hunk 2]
```diff
@@ -19,14 +19,19 @@ metadata:
...
+REPO_SCAN_INSTALL_DIR="${CLAUDE_CONFIG_DIR:-$HOME/.claude}/skills/repo-scan"
+trap 'rm -rf "$REPO_SCAN_TMP"' EXIT
+
+git clone --filter=blob:none --no-checkout \
+  https://github.com/haibindev/repo-scan.git "$REPO_SCAN_TMP/source"
+git -C "$REPO_SCAN_TMP/source" checkout --detach "$REPO_SCAN_COMMIT"
+
+# Review "$REPO_SCAN_TMP/source", then install tracked files without .git metadata
+mkdir -p "$REPO_SCAN_INSTALL_DIR"
+git -C "$REPO_SCAN_TMP/source" archive HEAD | tar -xf - -C "$REPO_SCAN_INSTALL_DIR"
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply skills/repo-scan/SKILL.md]: Fixed in 890340de. All locale blocks use set -euo pipefail and archive the exact pinned commit instead of HEAD.

[the remark this answers is not in the document]
Comment (dajiaohuang) [reply skills/repo-scan/SKILL.md]: Fixed in 890340de. Installation now requires typing install after reviewing the pinned source; EOF or any other response cancels without replacing the target.

[the remark this answers is not in the document]
Comment (dajiaohuang) [reply skills/repo-scan/SKILL.md]: Fixed in 890340de. The pinned archive is extracted into fresh staging, the existing target is moved aside only after confirmation, replacement is rollback-aware, and stale/.git files no longer survive. The exact flow was exercised against both fresh and stale targets.

File: tests/skills/repo-scan-install.test.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  path.join('docs', 'zh-CN', 'skills', 'repo-scan', 'SKILL.md'),
+  path.join('docs', 'ja-JP', 'skills', 'repo-scan', 'SKILL.md')
+];
+const pinnedCommit = '2742664ebcad1450c208eda0ae45d3c17fad5dd8';
+
+function installationBlock(relativePath) {
+  const source = fs.readFileSync(path.join(repoRoot, relativePath), 'utf8');
+  const match = source.match(/```bash\n([\s\S]*?)```/);
+  assert.ok(match, `${relativePath} must contain a bash installation block`);
+  return match[1];
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/skills/repo-scan-install.test.js]: Fixed in 890340de. The test now locates each locale-specific installation heading and restricts fence extraction to that section.

File: tests/skills/repo-scan-install.test.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+const blocks = skillFiles.map(installationBlock);
+for (const [index, block] of blocks.entries()) {
+  const relativePath = skillFiles[index];
+  assert.ok(block.includes(`REPO_SCAN_COMMIT=${pinnedCommit}`), `${relativePath} must pin the full commit SHA`);
+  assert.ok(block.includes('git clone --filter=blob:none --no-checkout'), `${relativePath} must clone before checkout`);
+  assert.ok(block.includes('checkout --detach "$REPO_SCAN_COMMIT"'), `${relativePath} must detach at the pin`);
+  assert.ok(block.includes('git -C "$REPO_SCAN_TMP/source" archive HEAD | tar -xf -'), `${relativePath} must install tracked files from the archive stream`);
+  assert.ok(block.includes('${CLAUDE_CONFIG_DIR:-$HOME/.claude}'), `${relativePath} must honor CLAUDE_CONFIG_DIR`);
+  assert.ok(!block.includes('cp -r .'), `${relativePath} must not copy .git metadata`);
+  assert.ok(!block.includes('git fetch --depth 1 origin 2742664\n'), `${relativePath} must not fetch the short SHA`);
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/skills/repo-scan-install.test.js:281]: Fixed in 890340de. The regression contract now requires fail-fast mode, cleanup, source review, exact-SHA archive, staging, default-deny confirmation, and confirmation-before-replacement.

File: tests/skills/repo-scan-install.test.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+for (const block of blocks.slice(1)) {
+  assert.strictEqual(block, blocks[0], 'translated installation commands must stay synchronized');
+}
+
+console.log(`  Passed: ${skillFiles.length}`);
+console.log('  Failed: 0');
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/skills/repo-scan-install.test.js]: Keeping these two summary lines intentionally: tests/run-all.js parses Passed: N and Failed: N from child output to compute aggregate totals. Removing or debug-gating them would recreate the reporting defect identified in this review.

File: skills/repo-scan/SKILL.md [hunk 6]
```diff
@@ -19,14 +19,44 @@ metadata:
...
-cp -r . ~/.claude/skills/repo-scan
+# Clone first so the pinned commit can be reviewed before installation
+set -euo pipefail
+
+REPO_SCAN_COMMIT=2742664ebcad1450c208eda0ae45d3c17fad5dd8
+REPO_SCAN_TMP="$(mktemp -d)"
+REPO_SCAN_INSTALL_DIR="${CLAUDE_CONFIG_DIR:-$HOME/.claude}/skills/repo-scan"
+REPO_SCAN_STAGE="$REPO_SCAN_TMP/install"
+REPO_SCAN_BACKUP=""
+trap 'rm -rf "$REPO_SCAN_TMP"' EXIT
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply skills/repo-scan/SKILL.md]: Fixed in 3cda4c4a. The install parent is created first, and mktemp now creates staging/backup under that parent so replacement and rollback stay on the target filesystem.

[the remark this answers is not in the document]
Comment (dajiaohuang) [reply skills/repo-scan/SKILL.md]: Fixed in 3cda4c4a. Cleanup is now conditional: a successful replacement or restoration removes the workspace, while a failed rollback preserves the previous installation and reports its exact backup path.

**Comment by dajiaohuang** [comment]
Addressed the review-level integration-test request in 3cda4c4a. The extracted installation block is now syntax-checked and executed hermetically for all three locale copies against a local Git fixture (no network), covering fresh install, stale-target replacement, forced replacement failure with successful rollback, and forced rollback failure with backup preservation. Focused result: Passed: 20, Failed: 0; skill/path/catalog validators and ESLint also pass.

File: skills/repo-scan/SKILL.md [hunk 7]
```diff
@@ -19,14 +19,53 @@ metadata:
...
+fi
+if ! mv -- "$REPO_SCAN_STAGE" "$REPO_SCAN_INSTALL_DIR"; then
+  if [ -e "$REPO_SCAN_BACKUP" ] || [ -L "$REPO_SCAN_BACKUP" ]; then
+    if ! mv -- "$REPO_SCAN_BACKUP" "$REPO_SCAN_INSTALL_DIR"; then
+      REPO_SCAN_KEEP_TMP=1
+      printf 'Replacement and rollback failed; previous installation preserved at %s\n' \
+        "$REPO_SCAN_BACKUP" >&2
+    fi
+  fi
+  exit 1
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply skills/repo-scan/SKILL.md:117]: Fixed in the latest branch commit. Replacement and rollback now hold an exclusive lock under the install parent, and rollback explicitly preserves the backup if the target is recreated. Hermetic coverage includes lock contention and target recreation; all three locale blocks stay identical.

File: tests/skills/repo-scan-install.test.js [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    const preservedBackup = path.join(installParent, workspaces[0], 'previous-install');
+    assert.strictEqual(
+      fs.readFileSync(path.join(preservedBackup, 'old-marker.txt'), 'utf8'),
+      'previous installation\n'
+    );
+    assert.match(result.stderr, /previous installation preserved at/);
+  } finally {
+    fs.rmSync(root, { recursive: true, force: true });
+  }
+}
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/skills/repo-scan-install.test.js:238]: Fixed in the latest branch commit. Fixture preparation, failure-mode selection, preserved-backup checks, and scenario assertions are now separate helpers; executeInstallation is limited to orchestration. The six-scenario integration matrix passes for all three locale blocks (26/26 total).

File: tests/skills/repo-scan-install.test.js [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    for (const scenario of ['fresh', 'existing', 'replacement-failure', 'rollback-failure']) {
+      executeInstallation(block, scenario);
+      passed++;
+    }
+  }
+} else {
+  console.log('  Integration coverage skipped on Windows without ECC_TEST_BASH');
+}
+
+console.log(`  Passed: ${passed}`);
```
[the remark this answers is not in the document]
Comment (dajiaohuang) [reply tests/skills/repo-scan-install.test.js:315]: No change needed for this finding: the reviewed branch already emits `console.log('  Failed: 0')` immediately after `Passed` in tests/skills/repo-scan-install.test.js. The focused run confirms both lines and reports Passed: 26 / Failed: 0.

**Comment by dajiaohuang** [comment]
Addressed review 4927969038 in the latest branch commit. The previous `target-conflict` shim failed before native `mv`, so it did not cover directory nesting. Replacement and rollback now use native `mv -T` when supported; the macOS/BSD fallback uses unique transaction names, detects a nested move, restores the moved tree to the transaction workspace, and fails without nesting stage/backup under the recreated target.

The hermetic matrix now executes native `mv` conflicts for both replacement and rollback, including forced no-`-T` portable fallback scenarios. Result: Passed: 35, Failed: 0. Skill/path/catalog validators, ESLint, and diff checks also pass.

**Comment by haelyra** [comment]
Superseded by merged installer hotfix #2784. Your repo-scan bootstrap repair was integrated, tested in the full cross-platform matrix, and preserved with co-author credit. Thank you.
